### PR TITLE
fix: code method on registration and 2fa

### PR DESCRIPTION
**Pull request by Benehiko**
This PR fixes:
1. 2FA incorrectly populates code method group
2. Registration type assertion on identity schema containing traits other than strings.

## Related issue(s)

https://github.com/ory-corp/cloud/issues/5280

## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [ ] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [ ] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [ ] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

File: selfservice/strategy/code/strategy_login.go [hunk 1]
```diff
@@ -87,6 +87,9 @@ func (s *Strategy) HandleLoginError(r *http.Request, f *login.Flow, body *update
 }
 
 func (s *Strategy) PopulateLoginMethod(r *http.Request, requestedAAL identity.AuthenticatorAssuranceLevel, lf *login.Flow) error {
+	if requestedAAL > identity.AuthenticatorAssuranceLevel1 {
+		return nil
+	}
```
Comment (zepatrik) [inline selfservice/strategy/code/strategy_login.go:92]: Does this mean code cannot be used as second factor, or it even cannot be used as a first one when there is a second one?

File: selfservice/strategy/code/strategy_login_test.go [hunk 2]
```diff
@@ -451,6 +451,48 @@ func TestLoginCodeStrategy(t *testing.T) {
...
+
+				// 3. Submit OTP
+				s = submitLogin(ctx, t, s, tc.isSPA, func(v *url.Values) {
+					v.Set("code", loginCode)
+				}, false, func(t *testing.T, s *state, body string, res *http.Response) {
+					if tc.isSPA {
+						require.EqualValues(t, http.StatusOK, res.StatusCode)
+					} else {
+						require.EqualValues(t, http.StatusOK, res.StatusCode)
+					}
```
Comment (zepatrik) [inline selfservice/strategy/code/strategy_login_test.go:488]:
```suggestion
					require.EqualValues(t, http.StatusOK, res.StatusCode)
```

File: selfservice/strategy/code/strategy_login_test.go [hunk 3]
```diff
@@ -451,6 +451,48 @@ func TestLoginCodeStrategy(t *testing.T) {
 				require.NotNil(t, va)
 				require.True(t, va.Verified)
 			})
+
+			t.Run("case=should not populate on 2FA request", func(t *testing.T) {
+				ctx := context.Background()
+
+				// enable webauthn 2FA method
+				conf.MustSet(ctx, fmt.Sprintf("%s.%s.enabled", config.ViperKeySelfServiceStrategyConfig, "webauthn"), true)
```
Comment (zepatrik) [inline selfservice/strategy/code/strategy_login_test.go:459]:
```suggestion
				conf.MustSet(ctx, strings.Join([]string{config.ViperKeySelfServiceStrategyConfig, "webauthn", "enabled"}, "."), true)
```

File: selfservice/strategy/code/strategy_registration.go [hunk 4]
```diff
@@ -253,7 +251,7 @@ func (s *Strategy) registrationVerifyCode(ctx context.Context, f *registration.F
 
 	// Step 2: Check if the flow traits match the identity traits
 	for _, n := range container.NewFromJSON("", node.DefaultGroup, p.Traits, "traits").Nodes {
-		if !strings.EqualFold(f.GetUI().GetNodes().Find(n.ID()).Attributes.GetValue().(string), n.Attributes.GetValue().(string)) {
+		if f.GetUI().GetNodes().Find(n.ID()).Attributes.GetValue() != n.Attributes.GetValue() {
```
Comment (zepatrik) [inline selfservice/strategy/code/strategy_registration.go:254]: What if we now have strings that have different casing? I think we'll need a type switch here.

**Review by zepatrik — CHANGES_REQUESTED** [review]

File: selfservice/strategy/code/strategy_login.go (hunk 1 above)
Comment (Benehiko) [reply selfservice/strategy/code/strategy_login.go:92]: code isn't used in 2FA flows anyway atm. Right now this is a First factor only implementation and the nodes shouldn't render on a 2FA screen at all!

File: selfservice/strategy/code/strategy_login_test.go (hunk 3 above)
Comment (Benehiko) [reply selfservice/strategy/code/strategy_login_test.go:459]: seems awfully complicated for a test all of a sudden :sweat_smile:

Comment (zepatrik) [reply selfservice/strategy/code/strategy_login_test.go:459]: I guess a bit easier is
```suggestion
				conf.MustSet(ctx, config.ViperKeySelfServiceStrategyConfig + ".webauthn.enabled"), true)
```

**Comment by Benehiko** [comment]
Only playwright mobile tests that are failing... rest are :green_circle:

**Comment by aeneasr** [comment]
pls rebase :)

**Comment by Benehiko** [comment]
> pls rebase :)

lol

**Review by jonas-jonas — APPROVED** [review]

**Review by alnr — APPROVED** [review]

**Review by aeneasr — APPROVED** [review]

**Pull request merged**